File: packages/core/src/schemas/criterion.ts
import { z } from "zod";

export const CriterionIdSchema = z.string().brand<"CriterionId">();
export const CriterionStatus = z.enum(["pending", "pass", "fail"]);
export const CriterionSource = z.enum(["ai", "user"]);

export const CriterionSchema = z.object({
  id: CriterionIdSchema,
  card_id: z.string(),
  text: z.string(),
  status: CriterionStatus.default("pending"),
  source: CriterionSource.default("ai"),
  evidence: z.string().nullable().default(null),
  execution_id: z.string().nullable().default(null),
  position: z.number().int().default(0),
  created_at: z.string(),
  updated_at: z.string(),
});

export const CreateCriterionSchema = z.object({
  text: z.string().min(1).max(2000),
});

export const UpdateCriterionSchema = z.object({
  text: z.string().min(1).max(2000).optional(),
  status: CriterionStatus.optional(),
});

export const ReorderCriteriaSchema = z.array(
  z.object({ id: CriterionIdSchema, position: z.number().int() })
);

File: rust/crates/core/src/types/criterion.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq)]
#[ts(export, export_to = "../../../../packages/dashboard/src/types/generated/")]
#[serde(rename_all = "lowercase")]
pub enum CriterionStatus {
    Pending,
    Pass,
    Fail,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq)]
#[ts(export, export_to = "../../../../packages/dashboard/src/types/generated/")]
#[serde(rename_all = "lowercase")]
pub enum CriterionSource {
    Ai,
    User,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../../packages/dashboard/src/types/generated/")]
pub struct Criterion {
    pub id: String,
    pub card_id: String,
    pub text: String,
    pub status: CriterionStatus,
    pub source: CriterionSource,
    pub evidence: Option<String>,
    pub execution_id: Option<String>,
    pub position: i64,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Deserialize, TS)]
#[ts(export, export_to = "../../../../packages/dashboard/src/types/generated/")]
pub struct CreateCriterion {
    pub text: String,
    pub source: CriterionSource,
}

#[derive(Debug, Clone, Default, Deserialize, TS)]
#[ts(export, export_to = "../../../../packages/dashboard/src/types/generated/")]
pub struct UpdateCriterion {
    pub text: Option<String>,
    pub status: Option<CriterionStatus>,
    pub evidence: Option<Option<String>>,
}
